substrate-archive/src/database.rs

```diff
@@ -272,3 +275,111 @@ impl Insert for Metadata {
 		.map_err(Into::into)
 	}
 }
+
+#[async_trait]
+impl Insert for Traces {
+	async fn insert(mut self, conn: &mut DbConn) -> DbReturn {
+		log::debug!("Inserting Trace Data");
+		let mut batch = Batch::new(
+			"state_tracing",
+			r#"
+			INSERT INTO "state_traces" (
+				block_num, hash, is_event, timestamp, duration, file, line, trace_id, trace_parent_id, target, name, traces
+			) VALUES
+			"#,
+			r#"
+			ON CONFLICT DO NOTHING
+			"#,
+		);
+
+		let block_num = self.block_num();
+		let hash = self.hash().to_vec();
+
+		for span in self.spans.into_iter() {
+			let id = i32::try_from(span.id.into_u64())?;
+			let parent_id: Option<i32> =
+				if let Some(id) = span.parent_id { Some(i32::try_from(id.into_u64())?) } else { None };
+			let overall_time: i64 = time_to_std(span.overall_time)?.as_nanos().try_into()?;
+			batch.reserve(12)?;
+			if batch.current_num_arguments() > 0 {
+				batch.append(",");
+			}
+			batch.append("(");
```

Comment:
This is hard to review, would be cool with a helper function/macro or something in the future for handle `bind` and `append`

Reply:
Yeah, definitely something that needs to be done and would offer lots of QoL improvements over the current batch-insertion strategy. I opened #211 to track this w/ a bit of history around it. Definitely redoing the batch-insertion logic is on my list of todo's though